File: Strivers-A2Z-DSA-Sheet/11. Binary Trees/1. Traversals/javascript/05. Postorder Traversal.js
/*
Question:
Given the root of a binary tree, return the postorder traversal of its nodes' values.

Approach:
- postorder traversal visits the left subtree first, followed by the right subtree, and then the root node.
- We can solve this problem recursively by following the postorder traversal order.

Complexity Analysis:
- Since we are visiting each node once, the time complexity of this approach is O(N), where N is the number of nodes in the binary tree.
- The space complexity is O(N) as we are using the call stack for recursion.

Code:
*/

function postorderTraversal(root) {
    const ans = [];
    solve(root, ans);
    return ans;
}

function solve(root, ans) {
    if (!root) {
        return;
    }
    solve(root.left, ans);
    solve(root.right, ans);
    ans.push(root.val);
}

File: Strivers-A2Z-DSA-Sheet/11. Binary Trees/1. Traversals/python3/05. Postorder Traversal.py
"""
Question:
Given the root of a binary tree, return the postorder traversal of its nodes' values.

Approach:
- postorder traversal visits the left subtree first, followed by the right subtree, and then the root node.
- We can solve this problem recursively by following the postorder traversal order.

Complexity Analysis:
- Since we are visiting each node once, the time complexity of this approach is O(N), where N is the number of nodes in the binary tree.
- The space complexity is O(N) as we are using the call stack for recursion.

Code:
"""
from typing import List, Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def postorderTraversal(root: Optional[TreeNode]) -> List[int]:
    ans = []
    solve(root, ans)
    return ans

def solve(root: Optional[TreeNode], ans: List[int]) -> None:
    if not root:
        return
    solve(root.left, ans)
    solve(root.right, ans)
    ans.append(root.val)
